lib/amazonka/src/Amazonka/Env/Hooks.hs

```diff
@@ -0,0 +1,600 @@
+
+-- | Like 'addHook', adds an unconditional hook, but it also captures
+-- the @'AWSRequest' a@ constraint. Useful for handling every AWS
+-- request type in a generic way.
+addAWSRequestHook :: (AWSRequest a, Typeable a) => Hook a -> Hook a -> Hook a
+addAWSRequestHook = addHook
+
+-- | 'addAWSRequestHook_' is 'addAWSRequestHook' but for 'Hook_'s.
+addAWSRequestHook_ :: (AWSRequest a, Typeable a) => Hook_ a -> Hook_ a -> Hook_ a
+addAWSRequestHook_ = addHook_
+
+-- | @addHookFor \@a newHook oldHook@ When @a@ and @b@ are the same
+-- type, run the given 'Hook a' after all others, otherwise only run
+-- the existing hooks.
+--
+-- @
+-- -- Example: Run @getObjectRequestHook@ on anything that is a @GetObjectRequest@:
+-- requestHook (addHookFor @GetObjectRequest getObjectRequestHook) :: Hooks -> Hooks
+-- @
+addHookFor ::
+  forall a b. (Typeable a, Typeable b) => Hook a -> Hook b -> Hook b
+addHookFor newHook oldHook = \env ->
+  oldHook env >=> case eqT @a @b of
+    Just Refl -> newHook env
+    Nothing -> pure
+
+-- | When @a@ and @b@ are the same type, run the given 'Hook_ a' after
+-- all other hooks have run.
+--
+-- @
+-- -- Example: Run @aSignedRequestHook@ on anything that is a @Signed GetObjectRequest@:
+-- requestHook (addHookFor_ @(Signed GetObjectRequest) aSignedRequestHook) :: Hooks -> Hooks
+-- @
+addHookFor_ ::
+  forall a b. (Typeable a, Typeable b) => Hook_ a -> Hook_ b -> Hook_ b
+addHookFor_ newHook oldHook = \env a -> do
+  oldHook env a
+  case eqT @a @b of
+    Just Refl -> newHook env a
+    Nothing -> pure ()
+
+-- | When @a@ and @b@ are the same type, do not call any more hooks.
+--
+-- @
+-- -- Example: Prevent any request hooks from running against a @PutObjectRequest@:
+-- requestHook (removeHooksFor @PutObjectRequest) :: Hooks -> Hooks
+-- @
+removeHooksFor :: forall a b. (Typeable a, Typeable b) => Hook b -> Hook b
+removeHooksFor oldHook = \env -> case eqT @a @b of
+  Just Refl -> pure
+  Nothing -> oldHook env
+
+-- | When @a@ and @b@ are the same type, do not call any more hooks.
+--
+-- @
+-- -- Example: Prevent any error hooks from running against errors caused by a @PutObjectRequest@:
+-- errorHook (removeHooksFor @(Request PutObjectRequest, Error)) :: Hooks -> Hooks
+-- @
+removeHooksFor_ :: forall a b. (Typeable a, Typeable b) => Hook_ b -> Hook_ b
+removeHooksFor_ oldHook = \env a -> case eqT @a @b of
+  Just Refl -> pure ()
+  Nothing -> oldHook env a
+
+-- | Run the wrapped hook unless the given 'Fold' or 'Traversal'
+-- matches the error. You will probably want to use this with the
+-- error matchers defined by each service binding, allowing you to
+-- selectively silence specific errors:
+--
+-- @
+-- -- Assuming `env :: Amazonka.Env` and `putRequest :: DynamoDB.PutRequest`,
+-- -- this silences a single type of error for a single call:
+-- send (env & #hooks %~ errorHook (silenceError DynamoDB._ConditionalCheckFailedException))
+-- @
+silenceError ::
```

Comment:
Need to think more about this — in its current form, will have to put it in twice: once for non-final errors and once for final errors. We should be able to do better than this.